README
 FeedTools was designed to be a simple XML feed parser, generator, and translator with a built-in
 caching system.

== Example
  require 'feed_tools'
  
  slashdot_feed = FeedTools::Feed.open('http://www.slashdot.org/index.rss')
  slashdot_feed.title
  => "Slashdot"
  slashdot_feed.description
  => "News for nerds, stuff that matters"
  slashdot_feed.link       
  => "http://slashdot.org/"
  slashdot_feed.items.first.find_node("slash:hitparade/text()").to_s
  => "43,37,28,23,11,3,1"

== Installation
 You can install FeedTools as a gem:
  gem install feedtools
  
 Or you can install it from the tarball or zip packages on the download page
 and then extract it to your vendors directory as you would with any other
 Ruby library.
 
 After installation, you will either need to run in non-caching mode or set
 up a caching mechanism.  The database feed cache system currently included
 with FeedTools is the most common caching method.  To set up the database
 feed cache, you will first need to create the appropriate database schema.
 Schema files for MySQL, PostgreSQL, and SQLite have been included, but the
 preferred method of creating the schema within the Rails environment is with
 a migration file.  A migration file has been supplied with FeedTools and can
 be found in the db directory.  Run
 <tt>script/generate migration add_feed_tools_tables</tt> and then copy and
 paste the contents of db/migration.rb into your new migration file.

CHANGELOG
=== FeedTools 0.2.28
  * fixed major bug with autodiscovery
=== FeedTools 0.2.27
  * now uses a new URI class
  * full support for IRIs
  * uses html5lib instead of HTree
  * schema changed to allow additional functionality
  * added better support for license information
  * added method to allow feeds and feed items to override lazy parsing
  * added configuration setting to override lazy parsing
  * added serialization feature to allow preparsing with FeedUpdater
  * added support for HTTP Basic authentication
  * fixed some issues with identifying the correct base uri
  * added config path method to the database feed cache
  * added default time to live configuration variable
  * fixed autodiscovery redirect loops
  * improved handling of entities
  * no longer causes problems with Mongrel's monkey patch
=== FeedTools 0.2.26
  * added support for retarded, ugly Radio redirects
  * improved autodiscovery support
  * fixed lazy evaluation issue
=== FeedTools 0.2.25
  * fixed self reference href bug
  * fixed primary link assignment heuristic
  * link rel's now default to 'alternate' for atom feeds
  * better handling of language attributes
  * better support for threaded applications
  * generated rss 2.0 now includes author, managingEditor, webMaster elements
  * improved handling of author elements
  * improved handling of http timeouts
  * fixed issue with relative paths in the location header on http redirection
  * monkey patches now work with Ruby 1.8.4's version of REXML
  * REXML monkey patches moved into their own file
=== FeedTools 0.2.24
  * fixed autodiscovery bugs
  * updated autodiscovery to handle relative uris
  * added per feed configuration
  * rewrote http retrieval code
  * supports authenticated http proxies
  * supports basic http auth
  * fixed relative url resolution issues
  * corrected db schema files to match migration file
  * fixed bug in the save method
  * fixed some major http bugs
  * updated dependancies
  * no longer causes problems with frozen rails
=== FeedTools 0.2.23
  * autodiscovery implemented
  * now knows a title from a hole in the ground
  * now resolves relative urls when possible
  * changed default table name to "cached_feeds" to avoid name collisions
  * schema now uses "href" instead of "url"
  * feed cache is set to nil by default now
  * both summary and content elements are generated now
  * now supports proxies
  * now supports internationalized domain names if libidn is installed
  * fixed bug with feed merging referencing a method that was refactored
  * no longer dies if uuidtools gem is missing but the UUID class is defined
  * updated timestamp handling and generation
  * added support for entry sorting on any feed item field
  * added support for disabling entry sorting entirely
  * fixed issue with itunes categories
  * fixed itunes subtitle/summary
  * fixed entry assignment bug
  * fixed issued/published variable name mix-up
  * added support for the payload module
  * added support for xhtml:div elements
  * dc:date now preempts pubDate
  * added better support for the scriptingNews format
  * now correctly strips out wrapper div elements from text constructs
  * fixed issue with some atom links being incorrectly identified as images
  * reorganized some of the helper modules
  * made some portions of url normalization case insensitive
  * fixed issue with filename handling on Windows
=== FeedTools 0.2.22
  * fixed another atom generation error
=== FeedTools 0.2.21
  * fixed atom generation error
  * added summary element to generated atom xml
  * added warning message for improperly set up database tables
=== FeedTools 0.2.20
  * fixed some atom namespace bugs
  * added a summary method to feed items
  * added some support for the apple-wallpapers elements
=== FeedTools 0.2.19
  * lousy encoding support (as opposed to none at all)
  * xml processing instruction now correctly prefixes generated feeds
  * attributes are escaped properly when generating feeds
  * uppercase html is no longer sanitized for not being in the whitelist
  * added alias method for assigning to entries
  * changed the xpath querying to be much, much more DRY
  * find_node and find_all_nodes are actually useful now
  * full case-insensitivity implemented for the xpath helper methods
  * fixed bug in tests where some assertion failures could affect other tests
  * fixed bug where the feed item author would sometimes be parsed incorrectly
  * fixed bug where the convertLineBreaks element would break feed entries
  * default (i.e. preferred) methods will be Atom-style instead of RSS-style
  * default feed output format changed to Atom 1.0
  * itunes namespace corrected
  * fixed images property when dealing with atom
  * fixed atom link property
  * improved timestamp handling
  * whitespace nodes now ignored by REXML
  * added option to disable timestamp estimation
  * added option to limit time-to-live to some upper maximum
  * enclosures included in feed generation
  * no longer uses the cache at all for file:/// urls
  * changed itunes:keywords to use commas
  * testing now excludes cache testing by default
  * more tests
=== FeedTools 0.2.18
  * no longer ever polls more often than once every 30 minutes
  * fixed overlooked improperly refactored enclosure code
  * fixed issue with inner_xml incorrectly handling xml comments
  * added helper modules
  * test cases now implemented using helpers
  * fixed issue with timeouts
  * fixed stack overflow while estimating timestamps
  * fixed some namespace issue with atom
  * added base64 decoding support
=== FeedTools 0.2.17
  * more fixes for timestamping of feed items
  * fixed nil bug in root_node, feed_type, feed_version, build_xml
  * build_xml method is now properly unit tested
  * fixed issue with setting the Host http header
  * namespaces are no longer hard-coded for feed generation
  * better url normalization for file: style urls
  * generated timestamps are now in UTC
=== FeedTools 0.2.16

lib/feed_tools.rb
#--
# Copyright (c) 2005 Robert Aman
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#++

if Object.const_defined?(:FEED_TOOLS_NAMESPACES)
  warn("FeedTools may have been loaded improperly.  This may be caused " +
    "by the presence of the RUBYOPT environment variable or by using " +
    "load instead of require.  This can also be caused by missing " +
    "the Iconv library, which is common on Windows.")
end

FEED_TOOLS_ENV = ENV['FEED_TOOLS_ENV'] ||
                 ENV['RAILS_ENV'] ||
                 'development' # :nodoc:

FEED_TOOLS_NAMESPACES = {
  "access" => "http://www.bloglines.com/about/specs/fac-1.0",
  "admin" => "http://webns.net/mvcb/",
  "ag" => "http://purl.org/rss/1.0/modules/aggregation/",
  "annotate" => "http://purl.org/rss/1.0/modules/annotate/",
  "atom10" => "http://www.w3.org/2005/Atom",
  "atom03" => "http://purl.org/atom/ns#",
  "atom-blog" => "http://purl.org/atom-blog/ns#",
  "audio" => "http://media.tangent.org/rss/1.0/",
  "bitTorrent" =>"http://www.reallysimplesyndication.com/bitTorrentRssModule",
  "blogChannel" => "http://backend.userland.com/blogChannelModule",
  "blogger" => "http://www.blogger.com/atom/ns#",
  "cc" => "http://web.resource.org/cc/",
  "creativeCommons" => "http://backend.userland.com/creativeCommonsRssModule",
  "co" => "http://purl.org/rss/1.0/modules/company",
  "content" => "http://purl.org/rss/1.0/modules/content/",
  "cp" => "http://my.theinfo.org/changed/1.0/rss/",
  "dc" => "http://purl.org/dc/elements/1.1/",
  "dcterms" => "http://purl.org/dc/terms/",
  "email" => "http://purl.org/rss/1.0/modules/email/",
  "ev" => "http://purl.org/rss/1.0/modules/event/",
  "icbm" => "http://postneo.com/icbm/",
  "image" => "http://purl.org/rss/1.0/modules/image/",
  "indexing" => "urn:atom-extension:indexing",
  "feedburner" => "http://rssnamespace.org/feedburner/ext/1.0",
  "foaf" => "http://xmlns.com/foaf/0.1/",
  "foo" => "http://hsivonen.iki.fi/FooML",
  "fm" => "http://freshmeat.net/rss/fm/",
  "gd" => "http://schemas.google.com/g/2005",
  "gr" => "http://www.google.com/schemas/reader/atom/",
  "itunes" => "http://www.itunes.com/dtds/podcast-1.0.dtd",
  "l" => "http://purl.org/rss/1.0/modules/link/",
  "media" => "http://search.yahoo.com/mrss",
  "p" => "http://purl.org/net/rss1.1/payload#",
  "pingback" => "http://madskills.com/public/xml/rss/module/pingback/",
  "prism" => "http://prismstandard.org/namespaces/1.2/basic/",
  "rdf" => "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
  "rdfs" => "http://www.w3.org/2000/01/rdf-schema#",
  "ref" => "http://purl.org/rss/1.0/modules/reference/",
  "reqv" => "http://purl.org/rss/1.0/modules/richequiv/",
  "rss09" => "http://my.netscape.com/rdf/simple/0.9/",
  "rss10" => "http://purl.org/rss/1.0/",
  "rss11" => "http://purl.org/net/rss1.1#",
  "rss20" => "http://backend.userland.com/rss2",
  "search" => "http://purl.org/rss/1.0/modules/search/",
  "slash" => "http://purl.org/rss/1.0/modules/slash/",
  "soap" => "http://schemas.xmlsoap.org/soap/envelope/",
  "ss" => "http://purl.org/rss/1.0/modules/servicestatus/",
  "str" => "http://hacks.benhammersley.com/rss/streaming/",
  "sub" => "http://purl.org/rss/1.0/modules/subscription/",
  "syn" => "http://purl.org/rss/1.0/modules/syndication/",
  "taxo" => "http://purl.org/rss/1.0/modules/taxonomy/",
  "thr" => "http://purl.org/rss/1.0/modules/threading/",
  "ti" => "http://purl.org/rss/1.0/modules/textinput/",
  "trackback" => "http://madskills.com/public/xml/rss/module/trackback/",
  "wfw" => "http://wellformedweb.org/CommentAPI/",
  "wiki" => "http://purl.org/rss/1.0/modules/wiki/",
  "xhtml" => "http://www.w3.org/1999/xhtml",
  "xml" => "http://www.w3.org/XML/1998/namespace"
}

$:.unshift(File.dirname(__FILE__))
$: << (File.dirname(__FILE__) + "/feed_tools/vendor")

begin
  require 'feed_tools/version'

  begin
    require 'iconv'
  rescue Object
    warn("The Iconv library does not appear to be installed properly.  " +
      "FeedTools cannot function properly without it.")
    raise
  end

  require 'rubygems'
  
  if !defined?(Builder)
    gem('builder', '>= 1.2.4')
    require 'builder'
  end
  
  if !defined?(ActiveRecord)
    gem('activerecord', '>= 1.11.1')
    require 'active_record'
  end

  # Preload optional libraries.
  begin
    require 'tidy'
  rescue Object
  end
  begin
    require 'idn'
  rescue Object
  end  

  require 'feed_tools/vendor/htree'
  require 'feed_tools/vendor/uri'

  require 'net/http'

# TODO: Not used yet, don't load since it'll only be a performance hit
#  require 'net/https'
#  require 'net/ftp'

  require 'rexml/document'

  require 'uri'
  require 'time'
  require 'cgi'
  require 'pp'
  require 'yaml'
  require 'base64'

  begin
    gem('uuidtools', '>= 0.1.2')
  rescue Gem::LoadError
    begin
      require 'uuidtools'
    rescue Object
      raise unless defined? UUID
    end
  end
  
  require 'feed_tools/monkey_patch'
  
  require 'feed_tools/feed'
  require 'feed_tools/feed_item'
  require 'feed_tools/feed_structures'
  require 'feed_tools/database_feed_cache'
  
  require 'feed_tools/helpers/html_helper'
  require 'feed_tools/helpers/xml_helper'
  require 'feed_tools/helpers/uri_helper'
rescue LoadError
  # ActiveSupport will very likely mess this up.  So drop a warn so that the
  # programmer can figure it out if things get wierd and unpredictable.
  warn("Unexpected LoadError, it is likely that you don't have one of the " +
    "libraries installed correctly.")
  raise
end

#= feed_tools.rb
#
# FeedTools was designed to be a simple XML feed parser, generator, and translator with a built-in
# caching system.
#
#== Example
#  slashdot_feed = FeedTools::Feed.open('http://www.slashdot.org/index.rss')
#  slashdot_feed.title
#  => "Slashdot"
#  slashdot_feed.description
#  => "News for nerds, stuff that matters"
#  slashdot_feed.link       
#  => "http://slashdot.org/"
#  slashdot_feed.items.first.find_node("slash:hitparade/text()").value
#  => "43,37,28,23,11,3,1"
module FeedTools
  @configurations = {}
  
  def FeedTools.load_configurations
    if @configurations.blank?
      # TODO: Load this from a config file.
      config_hash = {}
      @configurations = {
        :feed_cache => nil,
        :disable_update_from_remote => false,
        :proxy_address => nil,
        :proxy_port => nil,
        :proxy_user => nil,
        :proxy_password => nil,
        :auth_user => nil,
        :auth_password => nil,
        :auth_scheme => nil,
        :http_timeout => nil,
        :user_agent =>
          "FeedTools/#{FeedTools::FEED_TOOLS_VERSION::STRING} " + 
          "+http://www.sporkmonger.com/projects/feedtools/",
        :generator_name =>
          "FeedTools/#{FeedTools::FEED_TOOLS_VERSION::STRING}",
        :generator_href =>
          "http://www.sporkmonger.com/projects/feedtools/",
        :tidy_enabled => false,
        :tidy_options => {},
        :lazy_parsing_enabled => true,
        :serialization_enabled => false,
        :idn_enabled => true,
        :sanitization_enabled => true,
        :sanitize_with_nofollow => true,
        :always_strip_wrapper_elements => true,
        :timestamp_estimation_enabled => true,
        :url_normalization_enabled => true,
        :entry_sorting_property => "time",
        :strip_comment_count => false,
        :tab_spaces => 2,
        :max_ttl => 3.days.to_s,
        :default_ttl => 1.hour.to_s,
        :output_encoding => "utf-8"
      }.merge(config_hash)
    end
    return @configurations
  end
  
  # Resets configuration to a clean load
  def FeedTools.reset_configurations
    @configurations = nil
    FeedTools.load_configurations
  end
  
  # Returns the configuration hash for FeedTools
  def FeedTools.configurations
    if @configurations.blank?
      FeedTools.load_configurations()
    end
    return @configurations
  end
  
  # Sets the configuration hash for FeedTools
  def FeedTools.configurations=(new_configurations)
    @configurations = new_configurations
  end
  
  # Error raised when a feed cannot be retrieved    
  class FeedAccessError < StandardError
  end
  
  # Returns the current caching mechanism.
  #
  # Objects of this class must accept the following messages:
  #  id
  #  id=
  #  url
  #  url=
  #  title
  #  title=
  #  link
  #  link=
  #  feed_data
  #  feed_data=
  #  feed_data_type
  #  feed_data_type=
  #  etag
  #  etag=
  #  last_modified
  #  last_modified=
  #  save
  #
  # Additionally, the class itself must accept the following messages:
  #  find_by_id
  #  find_by_url
  #  initialize_cache
  #  connected?
  def FeedTools.feed_cache
    return nil if FeedTools.configurations[:feed_cache].blank?
    class_name = FeedTools.configurations[:feed_cache].to_s
    if @feed_cache.nil? || @feed_cache.to_s != class_name
      begin
        cache_class = eval(class_name)
        if cache_class.kind_of?(Class)
          @feed_cache = cache_class
          if @feed_cache.respond_to? :initialize_cache
            @feed_cache.initialize_cache
          end
          return cache_class
        else
          return nil
        end
      rescue
        return nil
      end
    else
      return @feed_cache
    end
  end
  
  # Returns true if FeedTools.feed_cache is not nil and a connection with
  # the cache has been successfully established.  Also returns false if an
  # error is raised while trying to determine the status of the cache.
  def FeedTools.feed_cache_connected?
    begin
      return false if FeedTools.feed_cache.nil?
      return FeedTools.feed_cache.connected?
    rescue
      return false
    end
  end    
  
  # Creates a merged "planet" feed from a set of urls.
  #
  # Options are:
  # * <tt>:multi_threaded</tt> - If set to true, feeds will
  #   be retrieved concurrently.  Not recommended when used
  #   in conjunction with the DatabaseFeedCache as it will
  #   open multiple connections to the database.
  def FeedTools.build_merged_feed(url_array, options = {})
    FeedTools::GenericHelper.validate_options([ :multi_threaded ],
                     options.keys)
    options = { :multi_threaded => false }.merge(options)
    warn("FeedTools.build_merged_feed is deprecated.")
    return nil if url_array.nil?
    merged_feed = FeedTools::Feed.new
    retrieved_feeds = []
    if options[:multi_threaded]
      feed_threads = []
      url_array.each do |feed_url|
        feed_threads << Thread.new do
          feed = Feed.open(feed_url)
          retrieved_feeds << feed
        end
      end
      feed_threads.each do |thread|
        thread.join
      end
    else
      url_array.each do |feed_url|
        feed = Feed.open(feed_url)
        retrieved_feeds << feed
      end
    end
    retrieved_feeds.each do |feed|
      merged_feed.entries = merged_feed.entries.concat(
        feed.entries.collect do |entry|
          new_entry = entry.dup
          new_entry.title = "#{feed.title}: #{entry.title}"
          new_entry
        end
      )
    end
    return merged_feed
  end
end

begin
  unless FeedTools.feed_cache.nil?
    FeedTools.feed_cache.initialize_cache
  end
rescue
end